### PR TITLE
PXC-4275: Make the SST errors more verbose

### PR DESCRIPTION
**Pull request by kamil-holubicki**
https://jira.percona.com/browse/PXC-4275

Problem:
To prevent infinite hangs of SST process in environment with unstable network, PXC introduced two configuration variables: to sst-initial-timeout and sst-idle-timeout.
When SST was terminated because stuck-prevention mechanism kicked-in, the error messages displayed were not informative enough.

Solution:
Improved SST script error messages.

### CODE REVIEW

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 1]
```diff
@@ -1371,19 +1378,39 @@ recv_data_from_donor_to_joiner()
...
-        wsrep_log_error "******************* FATAL ERROR ********************** "
+        wsrep_log_error "****************************************************** "
         exit 32
     fi
 
     if [[ ${RC[0]} -eq 124 ]]; then
         wsrep_log_error "******************* FATAL ERROR ********************** "
         wsrep_log_error "Possible timeout in receving first data from donor in gtid/keyring stage"
+        wsrep_log_error "This message is related to sst-initial-timeout/joiner-timeout configuration variable."
+        wsrep_log_error "It's current value is $WSREP_SST_JOINER_TIMEOUT seconds."
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]:
```suggestion
        wsrep_log_error "Its current value is $WSREP_SST_JOINER_TIMEOUT seconds."
```

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 2]
```diff
@@ -2054,19 +2081,37 @@ then
...
 
+        # Postpone script exit until all errors are reported
+        do_exit=0
         if [ ${RC[0]} -ne 0 ]; then
             wsrep_log_error "******************* FATAL ERROR ********************** "
+            # This is the only way, we can guess if it was the network connectivity problem.
+            # If socat gives up after -T seconds of stall, it exits with no error
+            # and pxb prints "OS errno 32 - Broken pipe" messages to its error log
+            # and then exits with code 1.
+            if cat "${DATA}/innobackup.backup.log" | grep -q "OS errno 32 - Broken pipe"; then
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: Instead of `cat` followed by `grep`, can we not just use `grep` directly on the file?

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 3]
```diff
@@ -2054,19 +2081,37 @@ then
...
+                wsrep_log_error "${XTRABACKUP_BIN} was not able to send data to the Joiner node."
+                wsrep_log_error "This error is most probably related to sst-idle-timeout configuration variable."
+                wsrep_log_error "It's current value is $WSREP_SST_IDLE_TIMEOUT seconds."
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]:
```suggestion
                wsrep_log_error "Its current value is $WSREP_SST_IDLE_TIMEOUT seconds."
```

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 4]
```diff
@@ -314,7 +314,14 @@ monitor_sst_progress() {
     fi
 
     if [[ ${current_timeout} -eq  ${timeout} ]]; then
+      wsrep_log_error "******************* FATAL ERROR ********************** "
       wsrep_log_error "Killing SST ($pid) with SIGKILL after stalling for ${current_timeout} seconds"
+      wsrep_log_error "This message is related to sst-idle-timeout configuration variable."
+      wsrep_log_error "It's current value is $WSREP_SST_IDLE_TIMEOUT seconds."
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]:
```suggestion
      wsrep_log_error "Its current value is $WSREP_SST_IDLE_TIMEOUT seconds."
```

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 5]
```diff
@@ -314,7 +314,14 @@ monitor_sst_progress() {
...
+      wsrep_log_error "PXC node didn't receive any data for last ${current_timeout} seconds during State Transfer (SST)."
+      wsrep_log_error "The most probable reason is broken network connectivity between Donor and Joiner (this) node."
+      wsrep_log_error "Check the network connection and restart the Joiner node."
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: Can we rephrase it to something like?

_The SST process on the joiner (this node) did not receive any data from the donor despite waiting for ${current_timeout} seconds (defined by the sst-idle-timeout variable). This could be due to a broken network connectivity between the donor and the joiner (this node). Consider checking the network connection and restarting the Joiner node._

@patrickbirch Any suggestions?

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 6]
```diff
@@ -1371,19 +1378,39 @@ recv_data_from_donor_to_joiner()
...
 
     if [[ ${RC[0]} -eq 124 ]]; then
         wsrep_log_error "******************* FATAL ERROR ********************** "
         wsrep_log_error "Possible timeout in receving first data from donor in gtid/keyring stage"
+        wsrep_log_error "This message is related to sst-initial-timeout/joiner-timeout configuration variable."
+        wsrep_log_error "It's current value is $WSREP_SST_JOINER_TIMEOUT seconds."
+        wsrep_log_error "This message means that SST transfer didn't start in $WSREP_SST_JOINER_TIMEOUT seconds"
+        wsrep_log_error "after requesting it."
+        wsrep_log_error "The most probable reason is broken network connectivity between Donor and Joiner (this) node."
+        wsrep_log_error "Check the network connection and restart the Joiner node."
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: Can we rephrase it to something like?

_The SST process on the donor did not start in the last ${WSREP_SST_JOINER_TIMEOUT} seconds (defined by the sst-initial-timeout variable) after requesting for the state transfer. This could be due to a broken network connectivity between the donor and the joiner (this node). Consider checking the network connection and restarting the Joiner node._

@patrickbirch Any suggestions?

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 7]
```diff
@@ -314,7 +314,14 @@ monitor_sst_progress() {
     fi
 
     if [[ ${current_timeout} -eq  ${timeout} ]]; then
+      wsrep_log_error "******************* FATAL ERROR ********************** "
       wsrep_log_error "Killing SST ($pid) with SIGKILL after stalling for ${current_timeout} seconds"
+      wsrep_log_error "This message is related to sst-idle-timeout configuration variable."
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: Should we also tell users that they can configure it in the [sst] section of the cnf file, if the current timeout is not sufficient for them?

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 8]
```diff
@@ -2054,19 +2081,37 @@ then
...
+            # and pxb prints "OS errno 32 - Broken pipe" messages to its error log
+            # and then exits with code 1.
+            if cat "${DATA}/innobackup.backup.log" | grep -q "OS errno 32 - Broken pipe"; then
+                wsrep_log_error "${XTRABACKUP_BIN} was not able to send data to the Joiner node."
+                wsrep_log_error "This error is most probably related to sst-idle-timeout configuration variable."
+                wsrep_log_error "It's current value is $WSREP_SST_IDLE_TIMEOUT seconds."
+                wsrep_log_error "PXC node didn't send any data for last $WSREP_SST_IDLE_TIMEOUT seconds during State Transfer (SST)."
+                wsrep_log_error "The most probable reason is broken network connectivity between Donor (this) and Joiner node."
+                wsrep_log_error "Check the network connection and restart the Joiner node."
+            fi
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: @patrickbirch Any suggestions?

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: scripts/wsrep_sst_xtrabackup-v2.sh (hunk 7 above)
Comment (kamil-holubicki) [reply scripts/wsrep_sst_xtrabackup-v2.sh]: I think saying that it is related to XYZ configuration variable is enough. It is straightforward that as it is configuration variable, it can be configured...

File: scripts/wsrep_sst_xtrabackup-v2.sh (hunk 2 above)
Comment (kamil-holubicki) [reply scripts/wsrep_sst_xtrabackup-v2.sh]: done

**Review by svetasmirnova — APPROVED** [review]

**Comment by imp3142** [comment]
LGTM

File: scripts/wsrep_sst_xtrabackup-v2.sh (hunk 5 above)
Comment (patrickbirch) [reply scripts/wsrep_sst_xtrabackup-v2.sh]: After waiting for the ${current_timeout} seconds (defined by the sst-idle-timeout variable), the SST process on the joiner (this node) has not received any data from the donor. This error could be caused by broken network connectivity between the donor and the joiner (this node). Check the network connection and restart the joiner node.

**Review by venkatesh-prasad-v — APPROVED** [review]
